src/v/cloud_storage/topic_recovery_service.cc

```diff
@@ -0,0 +1,594 @@
+    vlog(cst_log.debug, "creating topics");
+
+    result<void, recovery_error_ctx> result = outcome::success();
+    try {
+        auto create_topics_results = co_await create_topics(request);
+        for (const auto& r : create_topics_results) {
+            if (r.ec != cluster::errc::success) {
+                vlog(cst_log.warn, "topic creation failed: {}", r);
+                _recovery_status.erase(r.tp_ns);
+            } else {
+                vlog(
+                  cst_log.debug,
+                  "topic created: {}, expected downloads: {}",
+                  r,
+                  _recovery_status[r.tp_ns].pending_downloads);
+            }
+        }
+
+        co_await propagate_state(state::recovering_data);
+        start_download_bg_tracker();
+    } catch (const ss::timed_out_error& err) {
+        result = recovery_error_ctx::make(
+          fmt::format("failed to create topics: {}", err),
+          recovery_error_code::error_creating_topics);
+    }
+
+    if (result.has_error()) {
+        co_await propagate_state(state::inactive);
+    }
+    co_return result;
+}
+
+static cluster::topic_configuration make_topic_config(
+  const cloud_storage::topic_manifest& tm, const recovery_request& request) {
+    auto topic_config = tm.get_topic_config();
+    if (!topic_config.has_value()) {
+        vlog(
+          cst_log.warn,
+          "skipping topic creation for {}, missing values in manifest",
+          tm.get_manifest_path());
+    }
+
+    cluster::topic_configuration topic_to_create_cfg(
+      topic_config->tp_ns.ns,
+      topic_config->tp_ns.tp,
+      topic_config->partition_count,
+      topic_config->replication_factor);
+
+    auto& topic_properties = topic_to_create_cfg.properties;
+    auto manifest_props = topic_config->properties;
+
+    topic_properties.compression = manifest_props.compression;
+    topic_properties.cleanup_policy_bitflags
+      = manifest_props.cleanup_policy_bitflags;
+    topic_properties.compaction_strategy = manifest_props.compaction_strategy;
+
+    topic_properties.retention_duration = tristate<std::chrono::milliseconds>{
+      config::shard_local_cfg().cloud_storage_recovery_temporary_retention_ms};
+    topic_properties.retention_local_target_ms
+      = tristate<std::chrono::milliseconds>{
+        config::shard_local_cfg()
+          .cloud_storage_recovery_temporary_retention_ms};
+
+    topic_properties.segment_size = manifest_props.segment_size;
+    topic_properties.timestamp_type = manifest_props.timestamp_type;
+    topic_properties.shadow_indexing = model::shadow_indexing_mode::full;
+    topic_properties.recovery = true;
+
+    if (request.retention_bytes().has_value()) {
+        topic_properties.retention_local_target_bytes = tristate<size_t>{
+          request.retention_bytes()};
+        topic_properties.retention_bytes = tristate<size_t>{
+          request.retention_bytes()};
+    }
+
+    if (request.retention_ms().has_value()) {
```

Comment:
Shouldn't we be reading out of `manifest_props` for the total retention settings (both retention_bytes and retention_ms) by default?  e.g. if restoring a topic that had 2 week retention, we should restore it with 2 week retention.

Reply:
This was a mistake (mostly due to not considering the changes related to local and total settings). I have now changed to the code to:

1. set total bounds from the manifest, these remain unchanged throughout recovery
2. set local time based bound from the config value (looking to see if the default can be size based)
3. override local bounds if specified by user
4. reset local bounds to nullopt at the end of recovery.

Trying to see if the step 4 value of nullopt is good enough, or should we use something else.